File: src/shell.py
from .entity import Entity
from .networkstate import EntitySnapshot
from .renderprimitives import EntityRenderState, PolygonPrimitive
from .trail import Trail
from typing import TYPE_CHECKING
import math

if TYPE_CHECKING:
    from .player import Player

class Shell(Entity):
    def __init__(self, game, player, x_launch, y_launch, x_launch_vel, y_launch_vel, launch_time, size, damage, white_out):
        super().__init__(game)
        self._player = player
        self._x_launch = x_launch
        self._y_launch = y_launch
        self._x_launch_vel = x_launch_vel
        self._y_launch_vel = y_launch_vel
        self._launch_time = launch_time
        self._size = size
        self._damage = damage
        self._white_out = white_out
        
        self._x = x_launch
        self._y = y_launch
        
        self._trail = Trail(game, self._x, self._y)
        self._game.add_entity(self._trail)

    def draw(self):
        # Draw as tiny triangle
        if not self._game.get_interface(): return

        # White plain color
        # Interface doesn't expose raw GL.
        # We need to map local coords to screen coords.
        
        # Tri: (0, 0.018), (0.03, -0.018), (-0.03, -0.018)
        # These are in game units relative to self._x, self._y
        
        self.get_graphics().draw_world_polygon(
            [
                (self._x + 0.00, self._y + 0.018),
                (self._x + 0.03, self._y - 0.018),
                (self._x - 0.03, self._y - 0.018),
            ],
            (255, 255, 255),
        )

    def get_render_state(self):
        return EntityRenderState(
            entity_id=self.get_entity_id(),
            entity_type=self.get_entity_type(),
            primitives=(
                PolygonPrimitive(
                    points=(
                        (self._x + 0.00, self._y + 0.018),
                        (self._x + 0.03, self._y - 0.018),
                        (self._x - 0.03, self._y - 0.018),
                    ),
                    colour=(255, 255, 255),
                ),
            ),
            metadata={"size": self._size, "damage": self._damage, "white_out": self._white_out},
        )

    def build_network_snapshot(self):
        return EntitySnapshot(
            entity_id=-1 if self.get_entity_id() is None else self.get_entity_id(),
            entity_type=self.get_entity_type(),
            position=self.get_position(),
            payload={
                "size": self._size,
                "damage": self._damage,
                "white_out": self._white_out,
                "player_number": getattr(self._player, "_number", None),
            },
        )

    def update(self, time):
        old_x = self._x
        old_y = self._y
        
        time_since_launch = self._game.get_time() - self._launch_time
        
        # Trajectory physics: x = vt + x0, y = vt - 0.5gt^2 + y0 ?
        # C++: _y = timeSinceLaunch * (_yLaunchVel - 5.0f * timeSinceLaunch) + _yLaunch;
        # 5.0f * t * t effectively means 0.5 * g * t^2 where g=10.
        
        self._x = time_since_launch * self._x_launch_vel + self._x_launch
        self._y = time_since_launch * (self._y_launch_vel - 5.0 * time_since_launch) + self._y_launch
        
        landscape = self._game.get_landscape()
        landscape_width = landscape.get_landscape_width() if landscape else 10.0
        
        # Out of bounds
        if self._x > landscape_width or self._x < -landscape_width:
            self._trail.lay_trail(self._x, self._y)
            self._trail.set_inactive()
            if self._player:
                self._player.record_shot(self._x, self._y, -1)
            return False
            
        # Ground Collision
        if landscape:
            coll_res = landscape.ground_collision(old_x, old_y, self._x, self._y)
            if coll_res[0]:
                 cx, cy = coll_res[1], coll_res[2]
                 self._trail.lay_trail(cx, cy)
                 self.explode(cx, cy, -1)
                 return False

        self._trail.lay_trail(self._x, self._y)
        
        # Tank Collision
        players = self._game.get_players()
        for i, p in enumerate(players):
            if p and p.get_tank():
                if p.get_tank().intersect_tank(old_x, old_y, self._x, self._y):
                    self.explode(self._x, self._y, i)
                    return False
        
        return True

    def explode(self, x, y, hit_tank_idx):
        self._trail.set_inactive()
        
        self._game.explosion(x, y, self._size, self._damage, hit_tank_idx, 
                             7 if self._white_out else 1, self._white_out, self._player)
        
        if self._player:
            self._player.record_shot(x, y, hit_tank_idx)
            
        # Suicide implies deletion in update loop return false

File: groundfire-0.25/src/shell.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : shell.cc
//
//          By : Tom Russell
//
//        Date : 08-Sep-02
//
// Description : Handles the shell projectiles (i.e. standard gun projectile / 
//               nukes)
//
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include "shell.hh"
#include "game.hh"
#include "tank.hh"
#include "blast.hh"
#include "common.hh"
#include "soundentity.hh"
#include "trail.hh"
#include "landscape.hh"
#include "player.hh"

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cShell
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cShell::cShell
(
    cGame   * game,
    cPlayer * player,
    float xLaunch,
    float yLaunch,
    float xLaunchVel,
    float yLaunchVel,
    float launchTime,
    float size,    // The size of the blast created when the shell explodes
    float damage,  // Amount of damage a direct hit from the shell will do 
    bool  whiteOut // Will the shell cause a white flash when it explodes? 
)
: cEntity (game),
  _player (player),
  _xLaunchVel (xLaunchVel),
  _yLaunchVel (yLaunchVel),
  _launchTime (launchTime),
  _size (size),
  _damage (damage),
  _whiteOut (whiteOut)
{
    _x = _xLaunch = xLaunch;
    _y = _yLaunch = yLaunch;

    // Create a new trail object to leave behind the shell
    _trail = new cTrail (game, _x, _y);
    _game->addEntity (_trail);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cShell
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cShell::~cShell
(
)
{
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : draw
//
// Description : Draw the shell.
//
////////////////////////////////////////////////////////////////////////////////
void
cShell::draw
(
    void
)
{
    glLoadIdentity ();

    glTranslatef (_x, _y, -6.0f);

    glColor3f (1.0f, 1.0f, 1.0f);

    // Shells are currently drawn as tiny triangles (they look like dots)
    glBegin (GL_TRIANGLES);
    glVertex3f ( 0.00f,  0.018f, 0.0f);
    glVertex3f ( 0.03f, -0.018f, 0.0f);
    glVertex3f (-0.03f, -0.018f, 0.0f);
    glEnd ();
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : update
//
// Description : Update the shell
//
////////////////////////////////////////////////////////////////////////////////
bool
cShell::update
(
    float time
)
{
    float oldX = _x;
    float oldY = _y;

    float collisionX;
    float collisionY;

    // Calculate the new position of the shell using a formula based on the 
    // elapsed time since the shell was fired. The advantage of this is that it
    // is completely independent of the game's frame rate.

    // Get the number of seconds passed since the shell was created.
    float timeSinceLaunch = _game->getTime () - _launchTime;

    _x = timeSinceLaunch * _xLaunchVel + _xLaunch;
    _y = timeSinceLaunch * (_yLaunchVel - 5.0f * timeSinceLaunch) + _yLaunch;

    // If the shell has left the side of the screen.
    if (_x >  (_game->getLandscape ())->getLandscapeWidth () ||
        _x < -(_game->getLandscape ())->getLandscapeWidth ()) 
    {

        // Lay a final piece of trail and stop laying trail.
        _trail->layTrail (_x, _y);
        _trail->setInactive ();

        // Tell the player who fired the shell that it's left the screen 
        // (this is for the AI)
        _player->recordShot (_x, _y, -1);

        // shell has left the side of the screen so destroy it
        delete this;
        return (false);
    }

    // Have we hit the ground?
    if ((_game->getLandscape ())->groundCollision (oldX, oldY, _x, _y, 
                                                   &collisionX, 
                                                   &collisionY))
    {       
        _trail->layTrail (collisionX, collisionY);

        explode (collisionX, collisionY, -1);

        return (false);
    }

    _trail->layTrail (_x, _y);

    // Check for collision with a tank
    cPlayer ** players = _game->getPlayers ();

    for (int i = 0; i < 8 && players[i] != NULL; i++)
    {
        if (players[i]->getTank ()->intersectTank (oldX, oldY, _x, _y))
        {                      
            explode (_x, _y, i);

            return (false);
        }
    }
    
    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : explode
//
// Description : Explode the shell at the coordinates provided
//
////////////////////////////////////////////////////////////////////////////////
void
cShell::explode
(
    float x, 
    float y,
    int   hitTank // The tank to deal full damage to (i.e. the tank we hit)
)
{ 
    _trail->setInactive ();
    
    _game->explosion (x, y,
                      _size,
                      _damage,
                      hitTank, 
                      _whiteOut ? 7 : 1,
                      _whiteOut,
                      _player);

    // Tell the player where the shot landed (needed by the AI)
    _player->recordShot (x, y, hitTank);

    delete this;
}
